crates/block-producer/src/batch_builder/mod.rs

```diff
@@ -27,38 +29,48 @@ use crate::{
 /// of provers for proof generation. Proving is currently unimplemented and is instead simulated via
 /// the given proof time and failure rate.
 pub struct BatchBuilder {
-    pub batch_interval: Duration,
-    pub workers: NonZeroUsize,
-    /// Used to simulate batch proving by sleeping for a random duration selected from this range.
-    pub simulated_proof_time: Range<Duration>,
-    /// Simulated block failure rate as a percentage.
+    /// Represents all batch building workers.
     ///
-    /// Note: this _must_ be sign positive and less than 1.0.
-    pub failure_rate: f32,
+    /// This pool is always at maximum capacity. Idle workers will be in a [`std::future::Ready`]
+    /// state and are immedietely available for a new batch building job.
+    ///
+    /// See also: [`BatchBuilder::wait_for_available_worker`].
+    worker_pool: JoinSet<()>,
+    batch_interval: Duration,
     /// The batch prover to use.
     ///
     /// If not provided, a local batch prover is used.
     batch_prover: BatchProver,
+    /// Simulated block failure rate as a percentage.
```

Comment:
nit: I know it was already like this (assertion message mentioned this already) but it wouldn't be a percentage if it's within 0 and 1, no?

Reply:
You can have more than 100% though in this context it makes no sense. I agree its a bit overkill :)